File: drivers/python/boyodb/external.py
"""
External data source helpers for BoyoDB.

Query S3, URLs, HDFS, and files without importing data.
"""

from dataclasses import dataclass
from typing import Any, Dict, List, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from .client import Client


@dataclass
class S3Config:
    """S3 connection configuration."""
    region: Optional[str] = None
    endpoint: Optional[str] = None
    access_key: Optional[str] = None
    secret_key: Optional[str] = None
    session_token: Optional[str] = None
    path_style: bool = False


class ExternalTables:
    """
    Query external data sources without importing.

    Supports S3, HTTP URLs, HDFS, and local files.
    """

    def __init__(self, client: "Client"):
        self._client = client

    def query_s3(
        self,
        path: str,
        sql: str = "SELECT *",
        format: Optional[str] = None,
        config: Optional[S3Config] = None,
        limit: Optional[int] = None,
    ) -> List[Dict[str, Any]]:
        """
        Query data from S3.

        Args:
            path: S3 path (s3://bucket/path/file.parquet)
            sql: SQL query (use 'data' as table alias)
            format: File format (parquet, csv, json) - auto-detected if not specified
            config: S3 configuration
            limit: Optional row limit

        Returns:
            Query results

        Example:
            results = ext.query_s3(
                "s3://my-bucket/data/events.parquet",
                "SELECT event_type, COUNT(*) FROM data GROUP BY event_type"
            )
        """
        format_arg = f", '{format}'" if format else ""
        limit_clause = f" LIMIT {limit}" if limit else ""

        # Build the query using s3() table function
        full_sql = f"""
            SELECT * FROM (
                {sql.replace('FROM data', f"FROM s3('{path}'{format_arg})")}
            ) AS subq
            {limit_clause}
        """

        # If simple SELECT *, simplify
        if sql.strip().upper() == "SELECT *":
            full_sql = f"SELECT * FROM s3('{path}'{format_arg}){limit_clause}"

        return list(self._client.query(full_sql))

    def query_url(
        self,
        url: str,
        sql: str = "SELECT *",
        format: Optional[str] = None,
        headers: Optional[Dict[str, str]] = None,
        limit: Optional[int] = None,
    ) -> List[Dict[str, Any]]:
        """
        Query data from HTTP URL.

        Args:
            url: HTTP(S) URL
            sql: SQL query
            format: File format (csv, json, parquet)
            headers: Optional HTTP headers
            limit: Optional row limit

        Returns:
            Query results

        Example:
            results = ext.query_url(
                "https://example.com/data.csv",
                "SELECT * FROM data WHERE value > 100"
            )
        """
        format_arg = f", '{format}'" if format else ""
        limit_clause = f" LIMIT {limit}" if limit else ""

        if sql.strip().upper() == "SELECT *":
            full_sql = f"SELECT * FROM url('{url}'{format_arg}){limit_clause}"
        else:
            full_sql = f"""
                SELECT * FROM (
                    {sql.replace('FROM data', f"FROM url('{url}'{format_arg})")}
                ) AS subq
                {limit_clause}
            """

        return list(self._client.query(full_sql))

    def query_file(
        self,
        path: str,
        sql: str = "SELECT *",
        format: Optional[str] = None,
        limit: Optional[int] = None,
    ) -> List[Dict[str, Any]]:
        """
        Query data from local file.

        Args:
            path: Local file path
            sql: SQL query
            format: File format
            limit: Optional row limit

        Returns:
            Query results
        """
        format_arg = f", '{format}'" if format else ""
        limit_clause = f" LIMIT {limit}" if limit else ""

        if sql.strip().upper() == "SELECT *":
            full_sql = f"SELECT * FROM file('{path}'{format_arg}){limit_clause}"
        else:
            full_sql = f"""
                SELECT * FROM (
                    {sql.replace('FROM data', f"FROM file('{path}'{format_arg})")}
                ) AS subq
                {limit_clause}
            """

        return list(self._client.query(full_sql))

    def query_hdfs(
        self,
        path: str,
        sql: str = "SELECT *",
        format: Optional[str] = None,
        limit: Optional[int] = None,
    ) -> List[Dict[str, Any]]:
        """
        Query data from HDFS.

        Args:
            path: HDFS path (hdfs://namenode/path/file)
            sql: SQL query
            format: File format
            limit: Optional row limit

        Returns:
            Query results
        """
        format_arg = f", '{format}'" if format else ""
        limit_clause = f" LIMIT {limit}" if limit else ""

        if sql.strip().upper() == "SELECT *":
            full_sql = f"SELECT * FROM hdfs('{path}'{format_arg}){limit_clause}"
        else:
            full_sql = f"""
                SELECT * FROM (
                    {sql.replace('FROM data', f"FROM hdfs('{path}'{format_arg})")}
                ) AS subq
                {limit_clause}
            """

        return list(self._client.query(full_sql))

    def create_external_table(
        self,
        name: str,
        source_type: str,
        location: str,
        format: str,
        columns: List[Dict[str, str]],
        partition_columns: Optional[List[str]] = None,
    ) -> None:
        """
        Create a named external table.

        Args:
            name: Table name
            source_type: Source type (s3, url, hdfs, file)
            location: Source location
            format: File format
            columns: Column definitions [{"name": "id", "type": "INT64"}, ...]
            partition_columns: Optional partition columns

        Example:
            ext.create_external_table(
                "events_s3",
                "s3",
                "s3://bucket/events/",
                "parquet",
                [
                    {"name": "id", "type": "INT64"},
                    {"name": "event", "type": "STRING"},
                    {"name": "ts", "type": "TIMESTAMP"},
                ],
                partition_columns=["date"]
            )
        """
        col_defs = ", ".join(
            f"{c['name']} {c['type']}" for c in columns
        )
        partition_clause = ""
        if partition_columns:
            partition_clause = f" PARTITION BY ({', '.join(partition_columns)})"

        sql = f"""
            CREATE EXTERNAL TABLE {name} ({col_defs})
            STORED AS {format.upper()}
            LOCATION '{source_type}://{location.lstrip(source_type + "://")}'
            {partition_clause}
        """
        self._client.exec(sql)

    def drop_external_table(self, name: str) -> None:
        """Drop an external table."""
        self._client.exec(f"DROP EXTERNAL TABLE IF EXISTS {name}")


class VectorSearch:
    """
    Vector similarity search helpers.

    For AI/ML embedding search and nearest neighbor queries.
    """

    def __init__(self, client: "Client"):
        self._client = client

    def create_index(
        self,
        table: str,
        column: str,
        index_name: Optional[str] = None,
        index_type: str = "hnsw",
        distance_metric: str = "cosine",
        m: int = 16,
        ef_construction: int = 200,
    ) -> None:
        """
        Create a vector index.

        Args:
            table: Table name
            column: Vector column name
            index_name: Optional index name
            index_type: Index type (hnsw, ivfflat)
            distance_metric: Distance metric (cosine, l2, ip)
            m: HNSW M parameter
            ef_construction: HNSW ef_construction parameter
        """
        name = index_name or f"idx_{table.replace('.', '_')}_{column}_vec"

        sql = f"""
            CREATE INDEX {name} ON {table}
            USING {index_type} ({column} {distance_metric})
            WITH (m = {m}, ef_construction = {ef_construction})
        """
        self._client.exec(sql)

    def search(
        self,
        table: str,
        column: str,
        query_vector: List[float],
        k: int = 10,
        where: Optional[str] = None,
        ef_search: int = 100,
        return_columns: Optional[List[str]] = None,
    ) -> List[Dict[str, Any]]:
        """
        Find k nearest neighbors.

        Args:
            table: Table name
            column: Vector column
            query_vector: Query vector
            k: Number of results
            where: Optional filter condition
            ef_search: HNSW ef_search parameter
            return_columns: Columns to return (default: all)

        Returns:
            List of results with distance scores
        """
        vec_str = "[" + ",".join(str(v) for v in query_vector) + "]"
        cols = ", ".join(return_columns) if return_columns else "*"
        where_clause = f"WHERE {where}" if where else ""

        sql = f"""
            SELECT {cols}, {column} <=> '{vec_str}'::vector AS distance
            FROM {table}
            {where_clause}
            ORDER BY {column} <=> '{vec_str}'::vector
            LIMIT {k}
        """
        return list(self._client.query(sql))

    def hybrid_search(
        self,
        table: str,
        vector_column: str,
        text_column: str,
        query_vector: List[float],
        query_text: str,
        k: int = 10,
        vector_weight: float = 0.7,
        text_weight: float = 0.3,
    ) -> List[Dict[str, Any]]:
        """
        Hybrid vector + full-text search.

        Args:
            table: Table name
            vector_column: Vector column
            text_column: Text column for FTS
            query_vector: Query vector
            query_text: Query text
            k: Number of results
            vector_weight: Weight for vector similarity
            text_weight: Weight for text relevance

        Returns:
            List of results with combined scores
        """
        vec_str = "[" + ",".join(str(v) for v in query_vector) + "]"
        escaped_text = query_text.replace("'", "''")

        sql = f"""
            SELECT *,
                ({vector_weight} * (1.0 - ({vector_column} <=> '{vec_str}'::vector))) +
                ({text_weight} * ts_rank({text_column}_tsvector, plainto_tsquery('{escaped_text}')))
                AS hybrid_score
            FROM {table}
            WHERE {text_column}_tsvector @@ plainto_tsquery('{escaped_text}')
               OR {vector_column} <=> '{vec_str}'::vector < 0.5
            ORDER BY hybrid_score DESC
            LIMIT {k}
        """
        return list(self._client.query(sql))

File: drivers/nodejs/boyodb/external.js
/**
 * External data source helpers for BoyoDB Node.js driver.
 *
 * Query S3, URLs, HDFS, and files without importing data.
 */

/**
 * S3 connection configuration.
 * @typedef {Object} S3Config
 * @property {string} [region] - AWS region
 * @property {string} [endpoint] - Custom endpoint
 * @property {string} [accessKey] - Access key ID
 * @property {string} [secretKey] - Secret access key
 * @property {string} [sessionToken] - Session token
 * @property {boolean} [pathStyle=false] - Use path-style addressing
 */

/**
 * External tables helper for querying external data sources.
 *
 * @example
 * const ext = new ExternalTables(client);
 *
 * // Query S3
 * const results = await ext.queryS3(
 *   's3://my-bucket/data/events.parquet',
 *   'SELECT event_type, COUNT(*) as cnt FROM data GROUP BY event_type'
 * );
 *
 * // Query URL
 * const csvData = await ext.queryUrl(
 *   'https://example.com/data.csv',
 *   'SELECT * FROM data WHERE value > 100'
 * );
 */
class ExternalTables {
  /**
   * Create external tables helper.
   * @param {Client} client - BoyoDB client
   */
  constructor(client) {
    this._client = client;
  }

  /**
   * Query data from S3.
   * @param {string} path - S3 path (s3://bucket/path/file)
   * @param {string} [sql='SELECT *'] - SQL query (use 'data' as table alias)
   * @param {Object} [options={}] - Options
   * @param {string} [options.format] - File format (parquet, csv, json)
   * @param {S3Config} [options.config] - S3 configuration
   * @param {number} [options.limit] - Row limit
   * @returns {Promise<Object[]>}
   */
  async queryS3(path, sql = 'SELECT *', options = {}) {
    const formatArg = options.format ? `, '${options.format}'` : '';
    const limitClause = options.limit ? ` LIMIT ${options.limit}` : '';

    let fullSql;
    if (sql.trim().toUpperCase() === 'SELECT *') {
      fullSql = `SELECT * FROM s3('${path}'${formatArg})${limitClause}`;
    } else {
      fullSql = `
        SELECT * FROM (
          ${sql.replace(/FROM\s+data/gi, `FROM s3('${path}'${formatArg})`)}
        ) AS subq
        ${limitClause}
      `;
    }

    return await this._client.query(fullSql);
  }

  /**
   * Query data from HTTP URL.
   * @param {string} url - HTTP(S) URL
   * @param {string} [sql='SELECT *'] - SQL query
   * @param {Object} [options={}] - Options
   * @param {string} [options.format] - File format
   * @param {Object} [options.headers] - HTTP headers
   * @param {number} [options.limit] - Row limit
   * @returns {Promise<Object[]>}
   */
  async queryUrl(url, sql = 'SELECT *', options = {}) {
    const formatArg = options.format ? `, '${options.format}'` : '';
    const limitClause = options.limit ? ` LIMIT ${options.limit}` : '';

    let fullSql;
    if (sql.trim().toUpperCase() === 'SELECT *') {
      fullSql = `SELECT * FROM url('${url}'${formatArg})${limitClause}`;
    } else {
      fullSql = `
        SELECT * FROM (
          ${sql.replace(/FROM\s+data/gi, `FROM url('${url}'${formatArg})`)}
        ) AS subq
        ${limitClause}
      `;
    }

    return await this._client.query(fullSql);
  }

  /**
   * Query data from local file.
   * @param {string} path - Local file path
   * @param {string} [sql='SELECT *'] - SQL query
   * @param {Object} [options={}] - Options
   * @param {string} [options.format] - File format
   * @param {number} [options.limit] - Row limit
   * @returns {Promise<Object[]>}
   */
  async queryFile(path, sql = 'SELECT *', options = {}) {
    const formatArg = options.format ? `, '${options.format}'` : '';
    const limitClause = options.limit ? ` LIMIT ${options.limit}` : '';

    let fullSql;
    if (sql.trim().toUpperCase() === 'SELECT *') {
      fullSql = `SELECT * FROM file('${path}'${formatArg})${limitClause}`;
    } else {
      fullSql = `
        SELECT * FROM (
          ${sql.replace(/FROM\s+data/gi, `FROM file('${path}'${formatArg})`)}
        ) AS subq
        ${limitClause}
      `;
    }

    return await this._client.query(fullSql);
  }

  /**
   * Query data from HDFS.
   * @param {string} path - HDFS path (hdfs://namenode/path)
   * @param {string} [sql='SELECT *'] - SQL query
   * @param {Object} [options={}] - Options
   * @param {string} [options.format] - File format
   * @param {number} [options.limit] - Row limit
   * @returns {Promise<Object[]>}
   */
  async queryHdfs(path, sql = 'SELECT *', options = {}) {
    const formatArg = options.format ? `, '${options.format}'` : '';
    const limitClause = options.limit ? ` LIMIT ${options.limit}` : '';

    let fullSql;
    if (sql.trim().toUpperCase() === 'SELECT *') {
      fullSql = `SELECT * FROM hdfs('${path}'${formatArg})${limitClause}`;
    } else {
      fullSql = `
        SELECT * FROM (
          ${sql.replace(/FROM\s+data/gi, `FROM hdfs('${path}'${formatArg})`)}
        ) AS subq
        ${limitClause}
      `;
    }

    return await this._client.query(fullSql);
  }

  /**
   * Create a named external table.
   * @param {string} name - Table name
   * @param {Object} config - Table configuration
   * @param {string} config.sourceType - Source type (s3, url, hdfs, file)
   * @param {string} config.location - Source location
   * @param {string} config.format - File format
   * @param {Object[]} config.columns - Column definitions [{name, type}]
   * @param {string[]} [config.partitionColumns] - Partition columns
   * @returns {Promise<void>}
   */
  async createExternalTable(name, config) {
    const colDefs = config.columns
      .map(c => `${c.name} ${c.type}`)
      .join(', ');

    let partitionClause = '';
    if (config.partitionColumns && config.partitionColumns.length > 0) {
      partitionClause = ` PARTITION BY (${config.partitionColumns.join(', ')})`;
    }

    const sql = `
      CREATE EXTERNAL TABLE ${name} (${colDefs})
      STORED AS ${config.format.toUpperCase()}
      LOCATION '${config.sourceType}://${config.location.replace(new RegExp(`^${config.sourceType}://`), '')}'
      ${partitionClause}
    `;

    await this._client.exec(sql);
  }

  /**
   * Drop an external table.
   * @param {string} name - Table name
   * @returns {Promise<void>}
   */
  async dropExternalTable(name) {
    await this._client.exec(`DROP EXTERNAL TABLE IF EXISTS ${name}`);
  }
}

/**
 * Vector similarity search helper.
 *
 * @example
 * const vec = new VectorSearch(client);
 *
 * // Create index
 * await vec.createIndex('documents', 'embedding', {
 *   distanceMetric: 'cosine',
 *   m: 16,
 *   efConstruction: 200
 * });
 *
 * // Search
 * const results = await vec.search('documents', 'embedding', queryVector, {
 *   k: 10,
 *   where: "category = 'tech'"
 * });
 */
class VectorSearch {
  /**
   * Create vector search helper.
   * @param {Client} client - BoyoDB client
   */
  constructor(client) {
    this._client = client;
  }

  /**
   * Create a vector index.
   * @param {string} table - Table name
   * @param {string} column - Vector column
   * @param {Object} [options={}] - Index options
   * @param {string} [options.indexName] - Custom index name
   * @param {string} [options.indexType='hnsw'] - Index type (hnsw, ivfflat)
   * @param {string} [options.distanceMetric='cosine'] - Distance metric (cosine, l2, ip)
   * @param {number} [options.m=16] - HNSW M parameter
   * @param {number} [options.efConstruction=200] - HNSW ef_construction
   * @returns {Promise<void>}
   */
  async createIndex(table, column, options = {}) {
    const name = options.indexName || `idx_${table.replace(/\./g, '_')}_${column}_vec`;
    const indexType = options.indexType || 'hnsw';
    const distanceMetric = options.distanceMetric || 'cosine';
    const m = options.m || 16;
    const efConstruction = options.efConstruction || 200;

    const sql = `
      CREATE INDEX ${name} ON ${table}
      USING ${indexType} (${column} ${distanceMetric})
      WITH (m = ${m}, ef_construction = ${efConstruction})
    `;

    await this._client.exec(sql);
  }

  /**
   * Search for similar vectors.
   * @param {string} table - Table name
   * @param {string} column - Vector column
   * @param {number[]} queryVector - Query vector
   * @param {Object} [options={}] - Search options
   * @param {number} [options.k=10] - Number of results
   * @param {string} [options.where] - Filter condition
   * @param {number} [options.efSearch=100] - HNSW ef_search
   * @param {string[]} [options.returnColumns] - Columns to return
   * @returns {Promise<Object[]>}
   */
  async search(table, column, queryVector, options = {}) {
    const k = options.k || 10;
    const vecStr = `[${queryVector.join(',')}]`;
    const cols = options.returnColumns ? options.returnColumns.join(', ') : '*';
    const whereClause = options.where ? `WHERE ${options.where}` : '';

    const sql = `
      SELECT ${cols}, ${column} <=> '${vecStr}'::vector AS distance
      FROM ${table}
      ${whereClause}
      ORDER BY ${column} <=> '${vecStr}'::vector
      LIMIT ${k}
    `;

    return await this._client.query(sql);
  }

  /**
   * Hybrid vector + full-text search.
   * @param {string} table - Table name
   * @param {string} vectorColumn - Vector column
   * @param {string} textColumn - Text column for FTS
   * @param {number[]} queryVector - Query vector
   * @param {string} queryText - Query text
   * @param {Object} [options={}] - Search options
   * @param {number} [options.k=10] - Number of results
   * @param {number} [options.vectorWeight=0.7] - Vector similarity weight
   * @param {number} [options.textWeight=0.3] - Text relevance weight
   * @returns {Promise<Object[]>}
   */
  async hybridSearch(table, vectorColumn, textColumn, queryVector, queryText, options = {}) {
    const k = options.k || 10;
    const vectorWeight = options.vectorWeight || 0.7;
    const textWeight = options.textWeight || 0.3;
    const vecStr = `[${queryVector.join(',')}]`;
    const escapedText = queryText.replace(/'/g, "''");

    const sql = `
      SELECT *,
        (${vectorWeight} * (1.0 - (${vectorColumn} <=> '${vecStr}'::vector))) +
        (${textWeight} * ts_rank(${textColumn}_tsvector, plainto_tsquery('${escapedText}')))
        AS hybrid_score
      FROM ${table}
      WHERE ${textColumn}_tsvector @@ plainto_tsquery('${escapedText}')
         OR ${vectorColumn} <=> '${vecStr}'::vector < 0.5
      ORDER BY hybrid_score DESC
      LIMIT ${k}
    `;

    return await this._client.query(sql);
  }

  /**
   * Drop a vector index.
   * @param {string} indexName - Index name
   * @returns {Promise<void>}
   */
  async dropIndex(indexName) {
    await this._client.exec(`DROP INDEX IF EXISTS ${indexName}`);
  }
}

module.exports = {
  ExternalTables,
  VectorSearch,
};
